src/tslint/unifiedSignaturesRule.ts

```diff
@@ -24,241 +46,284 @@ export class Rule extends Lint.Rules.AbstractRule {
+		this.checkStatements(node.statements);
 		super.visitSourceFile(node);
 	}
 
 	visitModuleDeclaration(node: ts.ModuleDeclaration) {
 		const { body } = node;
 		if (body && body.kind === ts.SyntaxKind.ModuleBlock) {
-			this.visitStatements((body as ts.ModuleBlock).statements);
+			this.checkStatements((body as ts.ModuleBlock).statements);
 		}
 		super.visitModuleDeclaration(node);
 	}
 
 	visitInterfaceDeclaration(node: ts.InterfaceDeclaration): void {
-		const signatures = node.members.filter(m => m.kind === ts.SyntaxKind.CallSignature || m.kind === ts.SyntaxKind.MethodSignature) as
-			Array<ts.CallSignatureDeclaration | ts.MethodSignature>;
-		this.checkOverloads(signatures, node.typeParameters);
+		this.checkMembers(node.members, node.typeParameters);
 		super.visitInterfaceDeclaration(node);
 	}
 
 	visitClassDeclaration(node: ts.ClassDeclaration) {
-		this.visitMembers(node.members, node.typeParameters);
+		this.checkMembers(node.members, node.typeParameters);
 		super.visitClassDeclaration(node);
 	}
 
 	visitTypeLiteral(node: ts.TypeLiteralNode) {
-		this.visitMembers(node.members);
+		this.checkMembers(node.members);
 		super.visitTypeLiteral(node);
 	}
 
-	private visitStatements(statements: ts.Statement[]) {
-		this.checkOverloads(statements.filter(statement => statement.kind === ts.SyntaxKind.FunctionDeclaration) as ts.FunctionDeclaration[]);
-	}
-
-	private visitMembers(members: Array<ts.TypeElement | ts.ClassElement>, typeParameters?: ts.TypeParameterDeclaration[]) {
-		const signatures = members.filter(m =>
-			m.kind === ts.SyntaxKind.CallSignature || m.kind === ts.SyntaxKind.MethodSignature || m.kind === ts.SyntaxKind.MethodDeclaration) as
-			Array<ts.CallSignatureDeclaration | ts.MethodSignature | ts.MethodDeclaration>;
-		this.checkOverloads(signatures, typeParameters);
+	private checkStatements(statements: ts.Statement[]) {
+		this.checkOverloads(statements, (statement) => {
+			if (statement.kind === ts.SyntaxKind.FunctionDeclaration) {
+				const fn = statement as ts.FunctionDeclaration;
+				return fn.name && { signature: fn, key: fn.name.text };
+			} else {
+				return undefined;
+			}
+		});
 	}
 
-	private checkOverloads(signatures: ts.SignatureDeclaration[], typeParameters?: ts.TypeParameterDeclaration[]) {
-		const typeParametersSet = new Set((typeParameters || []).map(t => t.getText()));
-		for (const overloads of collectOverloads(signatures).values()) {
-			for (const [a, b] of pairs(overloads)) {
-				this.compareSignatures(a, b, type => typeParametersSet.has(type));
-			}
+	private checkMembers(members: Array<ts.TypeElement | ts.ClassElement>, typeParameters?: ts.TypeParameterDeclaration[]) {
+		this.checkOverloads(members, getOverloadName, typeParameters);
+		function getOverloadName(member: ts.TypeElement | ts.ClassElement) {
+			const key = getOverloadKey(member);
+			return key === undefined ? undefined : { signature: member as any as ts.SignatureDeclaration, key };
```

Comment:
1) `Type 'TypeElement | ClassElement' cannot be converted to type 'SignatureDeclaration'` because the compiler doesn't know that `getOverloadKey` checks this.
2) Key could be `""` (It shouldn't be, but compiler complains)